reana_server/scheduler.py

```diff
@@ -125,67 +135,106 @@ def get_consumers(self, Consumer, channel):
             )
         ]
 
-    def on_message(self, workflow_submission, message):
+    def _fail_workflow(self, workflow_id: str, logs: str = "") -> None:
+        self.workflow_status_publisher.publish_workflow_status(
+            workflow_id, status=RunStatus.failed.value, logs=logs,
+        )
+
+    def _retry_submission(self, workflow_id: str, workflow_submission: Dict) -> None:
+        retry_count = workflow_submission.get("retry_count", 0)
+        if retry_count >= REANA_SCHEDULER_REQUEUE_COUNT:
+            error_message = (
+                f"Workflow {workflow_submission['workflow_id_or_name']} failed to schedule after "
+                f"{retry_count} retries. Giving up."
+            )
+            logging.error(error_message)
+            self._fail_workflow(workflow_id, logs=error_message)
+        else:
+            workflow_submission["retry_count"] = retry_count + 1
+            workflow_submission["user_id"] = workflow_submission["user"]
+            del workflow_submission["user"]
+            current_workflow_submission_publisher.publish_workflow_submission(
+                **workflow_submission
+            )
+
+    def on_message(self, body, message):
         """On new workflow_submission event handler."""
-        workflow_submission = json.loads(workflow_submission)
-        logging.info("Received workflow: {}".format(workflow_submission))
+        workflow_submission = json.loads(body)
+        logging.info(f"Received workflow: {workflow_submission}")
+
+        workflow_submission_copy = workflow_submission.copy()
+
+        workflow_id = workflow_submission["workflow_id_or_name"]
         workflow_min_job_memory = workflow_submission.pop("min_job_memory", 0)
+
+        workflow_submission.pop("priority", None)
+        workflow_submission.pop("retry_count", None)
+
         if reana_ready(workflow_min_job_memory):
-            logging.info("Starting queued workflow: {}".format(workflow_submission))
-            workflow_submission.pop("priority", 0)
+            logging.info(f"Starting queued workflow: {workflow_id}")
             workflow_submission["status"] = "start"
+
+            retry = True
+            should_fail_on_error = True
+            started = False
+            error_message = ""
```

Comment:
Minor: I guess there is no point in keeping this variable here if the value never changes?

Reply:
I will still keep them as default values for visiblity.